e2store/src/era2.rs

```diff
@@ -0,0 +1,313 @@
+use ethportal_api::types::{execution::header::Header, state_trie::account_state::AccountState};
+
+use crate::{
+    e2store::{
+        stream::E2StoreStream,
+        types::{Entry, VersionEntry},
+    },
+    era1::HeaderEntry,
+};
+
+// <network-name>-<block-number>-<short-state-root>.era2
+//
+// era2 := Version | CompressedHeader | account*
+// account :=  CompressedAccount | CompressedStorage*
+// -----
+// Version                     = { type: 0x3265, data: nil }
+// CompressedHeader            = { type: 0x03,   data: snappyFramed(rlp(header)) }
+// CompressedAccount           = { type: 0x08,   data: snappyFramed(rlp(address_hash, rlp(nonce,
+// balance, storage_root, code_hash), raw_bytecode, storage_entry_count) }
+// CompressedStorage           = { type: 0x09,   data: snappyFramed(rlp(Vec<StorageItem {
+// storage_index_hash, value }>)) }
+// -----
+// CompressedStorage can have a max of 10k storage items, records must be filled before
+// creating a new one
+
+/// Represents an era2 `Era2` state snapshot.
+/// Unlike era1, not all fields will be stored in the struct, account's will be streamed from an
+/// iterator as needed.
+pub struct Era2 {
+    pub version: VersionEntry,
+    pub header: HeaderEntry,
+
+    /// e2store_stream, manages the interactions between the era2 state snapshot
+    e2store_stream: E2StoreStream,
+    storage_entries_left_to_decode: u16,
+}
+
+impl Era2 {
+    pub fn open(era2_path: &PathBuf) -> anyhow::Result<Self> {
+        let mut e2store_stream = E2StoreStream::open(era2_path)?;
+
+        let version = VersionEntry::try_from(&e2store_stream.next_entry()?)?;
+        let header = HeaderEntry::try_from(&e2store_stream.next_entry()?)?;
+
+        Ok(Self {
+            version,
+            header,
+            e2store_stream,
+            storage_entries_left_to_decode: 0,
+        })
+    }
+
+    pub fn create(era2_path: PathBuf, header: Header) -> anyhow::Result<Self> {
+        if era2_path.is_file() {
+            panic!(
+                "era2_path is not a directory, it is a file: {:?}",
+                era2_path
+            );
+        }
+        let era2_path = era2_path.join(format!(
+            "mainnet-{:010}-{}.era2",
+            header.number,
+            hex::encode(&header.state_root.as_slice()[..4])
+        ));
+        if era2_path.exists() {
+            panic!("era2 file already exists: {:?}", era2_path);
+        }
+        let mut e2store_stream = E2StoreStream::create(&era2_path)?;
+
+        let version = VersionEntry::default();
+        e2store_stream.append_entry(&version.clone().into())?;
+
+        let header = HeaderEntry { header };
+        e2store_stream.append_entry(&header.clone().try_into()?)?;
+
+        Ok(Self {
+            version,
+            header,
+            e2store_stream,
+            storage_entries_left_to_decode: 0,
+        })
+    }
+
+    #[allow(clippy::should_implement_trait)]
+    pub fn next(&mut self) -> anyhow::Result<Era2StreamEntry> {
```

Comment:
> How would we know that we finished reading the file

We get an error when we try to read an empty file

> nit: why not implement Iterator as clippy suggested?

Doesn't that require us to know how many entries there are? I don't think this is really possible, if you have a cleaver idea lets here it.

Reply:
> We get an error when we try to read an empty file

My question is not about reading empty file, it was about reading regular file, but we finished reading the content (and there are no entries left).
This is valid case, and instead of returning error that we can't easily distinguish from other errors (e.g. disk corruption, wrong encoding, etc), I think we should return `Option` to indicate that there is no more entries to read.

> Doesn't that require us to know how many entries there are? I don't think this is really possible, if you have a cleaver idea lets here it.

Iterators in general don't know how many entries they have (you might be confusing them with slices). Iterators can even be infinite. You can read more about them [here](https://doc.rust-lang.org/std/iter/index.html).

Reply:
Ok I implemented it, we will now panic if something fails, if we reach EOF we will return None